File: frontend/src/services/poll.ts
import { api } from "./api"
import { getVoterToken } from "@/utils/helpers"

export interface PollOption {
  id: string
  text: string
}

export interface Poll {
  id: string
  creatorId: string
  title: string
  description: string
  options: PollOption[]
  shareCode: string
  status: "active" | "closed"
  createdAt: string
  updatedAt: string
  expiresAt?: string
}

export interface PollResults {
  type: string
  pollId: string
  counts: Record<string, number>
  totalVotes: number
  timestamp: string
}

export interface CreatePollData {
  title: string
  description: string
  options: { text: string }[]
}

export const pollService = {
  // Creator (authenticated)
  async create(data: CreatePollData): Promise<Poll> {
    return api.post<Poll>("/polls", data, true)
  },

  async list(): Promise<Poll[]> {
    return api.get<Poll[]>("/polls", true)
  },

  async getById(id: string): Promise<Poll> {
    return api.get<Poll>(`/polls/${id}`, true)
  },

  async update(id: string, data: Partial<Pick<Poll, "title" | "description">>): Promise<Poll> {
    return api.patch<Poll>(`/polls/${id}`, data, true)
  },

  async delete(id: string): Promise<void> {
    return api.delete(`/polls/${id}`, true)
  },

  async close(id: string): Promise<Poll> {
    return api.post<Poll>(`/polls/${id}/close`, {}, true)
  },

  async getResults(id: string): Promise<PollResults> {
    return api.get<PollResults>(`/polls/${id}/results`, true)
  },

  // Public (no auth)
  async getPublic(shareCode: string): Promise<Poll> {
    return api.get<Poll>(`/public/polls/${shareCode}`)
  },

  async getPublicResults(shareCode: string): Promise<PollResults> {
    return api.get<PollResults>(`/public/polls/${shareCode}/results`)
  },

  async vote(pollId: string, optionId: string): Promise<PollResults> {
    const voterToken = getVoterToken()
    return api.post<PollResults>(`/polls/${pollId}/votes`, {
      optionId,
      voterToken,
    })
  },
}

File: backend/models/poll.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type PollOption struct {
	ID   string `bson:"id" json:"id"`
	Text string `bson:"text" json:"text"`
}

type PollStatus string

const (
	PollStatusActive PollStatus = "active"
	PollStatusClosed PollStatus = "closed"
)

type Poll struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	CreatorID   primitive.ObjectID `bson:"creatorId" json:"creatorId"`
	Title       string             `bson:"title" json:"title"`
	Description string             `bson:"description" json:"description"`
	Options     []PollOption       `bson:"options" json:"options"`
	ShareCode   string             `bson:"shareCode" json:"shareCode"`
	Status      PollStatus         `bson:"status" json:"status"`
	CreatedAt   time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt   time.Time          `bson:"updatedAt" json:"updatedAt"`
	ExpiresAt   *time.Time         `bson:"expiresAt,omitempty" json:"expiresAt,omitempty"`
}

type PollWithCounts struct {
	Poll
	TotalVotes int            `json:"totalVotes"`
	Counts     map[string]int `json:"counts"`
}
